coderd/schedule/autostop.go

```diff
@@ -0,0 +1,329 @@
+				// give a little leeway. This prevents skipped stop events
+				// because autostart perfectly lines up with autostop.
+				checkTime = checkTime.Add(restartRequirementBuffer)
+			}
+
+			// Get the next occurrence of the restart schedule.
+			autostop.MaxDeadline = userQuietHoursSchedule.Schedule.Next(checkTime)
+			if autostop.MaxDeadline.IsZero() {
+				return autostop, xerrors.New("could not find next occurrence of template restart requirement in user quiet hours schedule")
+			}
+		}
+	}
+
+	// If the workspace doesn't have a deadline or the max deadline is sooner
+	// than the workspace deadline, use the max deadline as the actual deadline.
+	if !autostop.MaxDeadline.IsZero() && (autostop.Deadline.IsZero() || autostop.MaxDeadline.Before(autostop.Deadline)) {
+		autostop.Deadline = autostop.MaxDeadline
+	}
+
+	if (!autostop.Deadline.IsZero() && autostop.Deadline.Before(now)) || (!autostop.MaxDeadline.IsZero() && autostop.MaxDeadline.Before(now)) {
+		// Something went wrong with the deadline calculation, so we should
+		// bail.
+		return autostop, xerrors.Errorf("deadline calculation error, computed deadline or max deadline is in the past for workspace build: deadline=%q maxDeadline=%q now=%q", autostop.Deadline, autostop.MaxDeadline, now)
+	}
+
+	return autostop, nil
+}
+
+// truncateMidnight truncates a time to midnight in the time object's timezone.
+// t.Truncate(24 * time.Hour) truncates based on the internal time and doesn't
+// factor daylight savings properly.
+//
+// See: https://github.com/golang/go/issues/10894
+func truncateMidnight(t time.Time) time.Time {
+	yy, mm, dd := t.Date()
+	return time.Date(yy, mm, dd, 0, 0, 0, 0, t.Location())
+}
+
+// nextDayMidnight returns the next midnight in the time object's timezone.
+func nextDayMidnight(t time.Time) time.Time {
+	yy, mm, dd := t.Date()
+	// time.Date will correctly normalize the date if it's past the end of the
+	// month. E.g. October 32nd will be November 1st.
+	dd++
+	return time.Date(yy, mm, dd, 0, 0, 0, 0, t.Location())
+}
+
+// WeeksSinceEpoch gets the weeks since the epoch for a given time. This is a
+// 0-indexed number of weeks since the epoch (Monday).
+//
+// The timezone embedded in the time object is used to determine the epoch.
+func WeeksSinceEpoch(now time.Time) (int64, error) {
+	epoch := TemplateRestartRequirementEpoch(now.Location())
+	if now.Before(epoch) {
+		return 0, xerrors.New("coder server system clock is incorrect, cannot calculate template restart requirement")
+	}
+
+	// This calculation needs to be done using YearDay, as dividing by the
+	// amount of hours is impacted by daylight savings. Even though daylight
+	// savings is usually only an hour difference, this calculation is used to
+	// get the current week number and could result in an entire week getting
+	// skipped if the calculation is off by an hour.
+	//
+	// Old naive algorithm: weeksSinceEpoch := int64(since.Hours() / (24 * 7))
```

Comment:
Daylight savings would cause the old algorithm to be off by the daylight savings offset after a single DST event has passed but not the second DST event in the timezone during that year.

The new algorithm operates on dates only and doesn't touch hours/minutes, so it avoids this issue. There are heaps of tests for this algorithm in `autostop_test.go`, it's tested in a bunch of popular timezones for the next 5200 weeks (aka. 100 years).